### PR TITLE
Adds initial version of ERC20Receiver extension.

### PR DESCRIPTION
**Pull request by steven2308**
# Description

Implementing IERC20Receiver extension.
It needs to use a pull mechanism for incoming transfers in order to identify the tokenId. 

# Checklist

- [x] Verified code additions
- [x] Updated NatSpec comments (if applicable)
- [x] Regenerated docs
- [x] Ran prettier
- [x] Added tests to fully cover the changes

---

### Detailed summary
- Updated the version of `prettier-plugin-solidity` to `1.1.3`
- Added `IERC20Holder` interface and its implementation in `RMRKERC20HolderMock` contract
- Added events and methods for transferring ERC-20 tokens in `IERC20Holder` interface

> The following files were skipped due to too many changes: `docs/RMRK/extension/erc20Holder/IERC20Holder.md`, `docs/RMRK/extension/erc20Holder/ERC20Holder.md`, `contracts/RMRK/extension/erc20Holder/ERC20Holder.sol`, `test/extensions/erc20holder.ts`, `docs/mocks/extensions/erc20holder/RMRKERC20HolderMock.md`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:32]: I'm thinking of adding a `bytes data` here, for advanced usage. Same on the transferTo method.

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:1]: All of the comments are out of order. When developing their solution, it might take a lot of work to parse for someone referring to the source code.

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:10]: I'd move this to the Errors library. 🤔

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:13]: If I'm assuming correctly, this is
```suggestion
    mapping(uint256 tokenId => mapping(address erc20address => uint256 balance)) private balances;
```
right?

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:32]: Yeah, I think it makes sense to include it 👌

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:47]: The actual transfer is still missing, right?

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:65]: Should we validate the allowance before attempting the transfer? This way we can return our own error, but it also increases contract size..

I think this bypasses the issue of users draining each other 👌

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:1]: It took me a bit to understand this hehe. Fixed, thanks!

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:26]: I think these comments need to be changed as the function transfers the tokens from the token and not to it 🤔 

```suggestion
     * @notice Transfer ERC-20 tokens from a specific token.
```

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:27]: I don't think the approval is needed, since the tokens are owned by this smart contract.

```suggestion

```

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:28]: I think that the balance is transferred from the smart contract as the tokens are owned by the NFT, right?

```suggestion
     * @dev The balance MUST be transferred from this smart contract.
```

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:101]:
```suggestion
     * @param data Additional data with no specified format, to allow for custom logic
```

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:117]:
```suggestion
     * @param data Additional data with no specified format, to allow for custom logic
```

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:133]:
```suggestion
     * @param data Additional data with no specified format, to allow for custom logic
```

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:149]:
```suggestion
     * @param data Additional data with no specified format, to allow for custom logic
```

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:35]: Should this also accept `destinationId` to be able to transfer the tokens from one NFT to another?

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:25]: Would you mind removing `.`s from `@param` tags? There are a lot to create suggestions for all of them 😅

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:26]: I think you are confused here. Since ERC20 has no onERC20Received method, this method will pull the tokens into the contract, to a specific token.

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:27]: On the same not of the previous comment. The contract will pull them from user balances, so it needs to be approved.

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:28]: Nope, same as previous 2 comments

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:35]: Did not think of this actually, it makes it more complex. But would be good to have. 
I don't like that we need an additional flag to know if destination is an NFT or not, since Id could be zero.  
I would prefer keeping it simply, WDTY?

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:82]: About validating allowance, it feels an overkill, it will also spend more gas every time so I'r rather not do it,

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:10]: I'm more an more against this. This could be a nice stand alone, I feel like we should keep errors for extensions only on extensions :thinking:

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol]: Yes, prettier was removing them.
I upgrade it and now it's fine.

**Comment by ThunderDeliverer** [comment]
@steven2308 would you mind rebasing to the current `dev`? I just rebased it after the merge with `master`, so that we won't have merge conflicts there.

**Comment by steven2308** [comment]
> @steven2308 would you mind rebasing to the current `dev`? I just rebased it after the merge with `master`, so that we won't have merge conflicts there.

Done. I did merge instead. Will do squash and merge for this when approved.

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:36]: I get the simplicity. Just thinking about the most usability? We've been dealing with simplicity vs. usability for all of the proposals we published..

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:27]: All this time I was thinking these comments were `transferERC20ToToken`, sorry :grimacing: 
You were right on each of them

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:35]: I've spent some time thinking about it and found a blocker. Remember we cannot push ERC20 tokens as there is no onERC20Received method. 
However, I see an alternative which would be to transfer ERC20s between NFTs of the same collection, it's more of a util, will add a method and let's see if it's worth it.

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:48]:
```suggestion
     * @notice Transfer ERC-20 tokens from a specific token.
```

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:51]:
```suggestion
     * @param erc20Contract The address of the ERC-20 smart contract
```

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:52]:
```suggestion
     * @param tokenId The ID of the token to transfer the ERC-20 tokens from
```

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:65]:
```suggestion
     * @notice Transfer ERC-20 tokens to a specific token.
```

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:66]:
```suggestion
     * @dev The ERC-20 smart contract must have approval for this contract to transfer the ERC-20 tokens.
```

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:67]:
```suggestion
     * @dev The balance MUST be transferred from the `msg.sender`.
```

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:69]:
```suggestion
     * @param erc20Contract The address of the ERC-20 smart contract
     * @param tokenId The ID of the token to transfer ERC-20 tokens to
```

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:81]:
```suggestion
     * @notice Transfer ERC-20 tokens from one token to another one within the same collection.
```

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:82]:
```suggestion
     * @dev ERC-20 tokens are only transferred internally; they never leave this smart contract.
```

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:84]:
```suggestion
     * @param erc20Contract The address of the ERC-20 smart contract
```

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (ThunderDeliverer) [inline contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:86]:
```suggestion
     * @param toTokenId The ID of the token to transfer ERC-20 tokens from
     * @param toTokenId The ID of the token to transfer ERC-20 tokens to
```

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:97]: I thought it was better than nothing, but I think I found a way, will give it a try. Maybe we can discuss it before,

File: contracts/RMRK/extension/erc20Holder/IERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/IERC20Holder.sol:27]: Feels more reasonable to me too. Although I can't explain why :sweat_drops: 
Will change.

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:35]: @ThunderDeliverer  I could not find a way to do this in a secure way.
I evaluated 2 options, say token 1 from collection A wants to send ERC20 tokens X to token 2 from collection B. Token A is owned by User.

Option A. Pulling.
1. User  needs to approve B to mange it's token. So the user calls `A.approve(B, 1)`.
2. Owner calls `B.receiveERC20FromToken(fromCollection: A, fromToken: 1, toToken 2, amount: x).`  
3. B would in turn call `A.transferERC20FromToken(erc20: X, fromToken: 1, toCollection: B, amount: x)`
4. B would include the amount x in balances held by 2.

This has a flaw, after step 1, anyone could do step 2 and transfer to theirs. Unless it checks that msg.sender owns A.1. Which could be done, but it's scary to say the least. A lot of trust has to be put on the receiver (B).

Also, the fact that you have to approve the destination contract is very disturbing, it would still be your fault but it makes it easier to fall for scams IMO.

Option B. Pushing (usually terrible idea, but let's see).
1.  User calls `A.transferERC20FromToken(erc20: X, fromToken: 1, toCollection: B, toTokenId: 2, amount: x)`
2.  A, after doing the transfer, notifies B that the funds it received where for token 2. 

The problem here is more obvious, A can lie.

Option C. Pushing V2
1.  User calls `A.transferERC20FromToken(erc20: X, fromToken: 1, toCollection: B, toTokenId: 2, amount: x)`
2. A approves B to spend X:  `X.approve(spender: B, amount: x)`
3. A calls `B.transferERC20ToToken(erc20: X, tokenId: 2, amount: x, from: A)` Notices that we need to add `from`, so msg.sender is not necessarily used.
4. A reduces balance of X from 1.

The problem here is that anyone could call step 3 with a different from that has some pending approval. 

I don't like option A, and would never do B or C. I consider the safest is to simply not do it, and if the transfer between tokens method makes it look limiting, I would simply remove it and have sources and destination be only addresses, not NFTs.

If you want to give it a try go ahead, but I'd rather have us focused on other priorities, this is perfectly good enough for the use case I've described in the past.

Comment (ThunderDeliverer) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:35]: I'll leave my thoughts on the possible approach and we can either consider them or remove the ability to transfer ERC20s between tokens.

The ERC20 smart contract considers smart contract A the owner of the tokens, and the User is only the owner of the NFT residing in A.
I'd add 2 functions; `sendFromTokenToToken` and `receiveFromToken` (names applicable to change 😅 ).
So, the User calls `sendFromTokenToToken` and the following happens (in the same transaction):
1. A calls ERC20 contract to grant the allowance for B
2. A calls B telling it how many tokens to transfer to a token
3. B transfers the ERC20s to token (the NFT to receive must exist)

I think this is it. No danger of another token to steal the ERC20s. The only danger is the possibility of malicious B, but this is a permanent vulnerability that can't be removed 🤔 

PS.: Also, the EIP wording will be a challenge as it can get confusing really fast 😂

Comment (steven2308) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:35]: Could work, wanna give it a try?

**Comment by ThunderDeliverer** [comment]
I think there is just a bit of open stuff left here. Do you want to do it @steven2308 so we can merge this and proceed with the proposal, or should I do it?

**Comment by steven2308** [comment]
> I think there is just a bit of open stuff left here. Do you want to do it @steven2308 so we can merge this and proceed with the proposal, or should I do it?

I was waiting for you to add a proposed implementation on NFT to NFT transfer. We could merge like this also and start from there.
In any case, let's not do a proposal just yet.

File: .env.example
Comment (ThunderDeliverer) [inline .env.example]: I'd keep the example here:

```suggestion
SEPOLIA_URL=https://eth-sepolia.g.alchemy.com/v2/<YOUR ALCHEMY KEY>
```

File: .env.example
Comment (ThunderDeliverer) [inline .env.example]: Same here:
```suggestion
ETHEREUM_URL=https://eth-mainnet.g.alchemy.com/v2/<YOUR ALCHEMY KEY>
```

File: hardhat.config.ts
Comment (ThunderDeliverer) [inline hardhat.config.ts]: Why did you remove `sepolia`?

**Review by ThunderDeliverer — CHANGES_REQUESTED** [review]

File: contracts/RMRK/extension/erc20Holder/ERC20Holder.sol
Comment (ThunderDeliverer) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:13]: Prettier removed them again 😦

Comment (ThunderDeliverer) [reply contracts/RMRK/extension/erc20Holder/ERC20Holder.sol:13]:
```suggestion
    mapping(uint256 tokenId => mapping(address erc20Address => uint256 balance)) private _balances;
```

**Review by ThunderDeliverer — APPROVED** [review]

File: hardhat.config.ts
Comment (steven2308) [reply hardhat.config.ts]: I think I just moved it down.

Comment (ThunderDeliverer) [reply hardhat.config.ts]: For some reason, the only change in diff was removing it, but I checked the file, and it is there. We are good to go! 🚀

**Pull request merged**
